lib/connect.js

```diff
@@ -234,29 +244,39 @@ const getTasks = (model, setupListeners) => {
         setupListeners(mongoClient);
       }
 
-      mongoClient.connect((err, _client) => {
-        ctx(err);
-
-        if (err) {
-          if (tunnel) {
-            debug('data-service connection error, shutting down ssh tunnel');
-            tunnel.close();
-          }
-
-          return cb(err);
+      /** @type {Promise<never>} */
+      const waitForTunnelError = new Promise(async(_, reject) => {
+        if (tunnel) {
+          const [error] = await once(tunnel, 'error');
+          reject(error);
         }
+      });
 
-        client = _client;
-
+      const closeTunnelOnError = async() => {
         if (tunnel) {
-          client.on('close', () => {
-            debug('data-service disconnected. shutting down ssh tunnel');
-            tunnel.close();
-          });
+          debug('data-service connection error, shutting down ssh tunnel');
+          try {
+            await tunnel.close();
+            debug('ssh tunnel stopped');
+          } catch (err) {
+            debug('ssh tunnel stopped with error: %s', err.message);
+          }
         }
+      };
 
-        cb(null, { url: model.driverUrlWithSsh, options: validOptions });
-      });
+      try {
+        const _client = await Promise.race([
```

Comment:
isn't this: `const [_client]`? is there a test that verifies if we return a client and a tunnel?

Reply:
`Promise.race` returns only one value (it races a bunch of them against each other, right?), so it's not an array.

We have a whole suite of tests for `connect` method and they are still passing, so `client` is returned correctly, this implementation doesn't change current behavior. I haven't added any tests for returning `tunnel` because I initially thought that this will require setting up a proper tunnel server to test the whole thing and it seemed like too much for a trivial test like that, but I can mock a bunch of things and write one for returning `tunnel`, yeah

Reply:
Nice! makes a lot of sense :)